neo/Consensus/ConsensusContext.cs

```diff
@@ -203,19 +205,84 @@ private void SignPayload(ConsensusPayload payload)
+        {
+            var blockSize =
+                // BlockBase
+                sizeof(uint) +       //Version
+                UInt256.Length +     //PrevHash
+                UInt256.Length +     //MerkleRoot
+                sizeof(ulong) +      //Timestamp
+                sizeof(uint) +       //Index
+                UInt160.Length +     //NextConsensus
+                1 +                  //
+                _witnessSize;        //Witness
+
+            blockSize +=
+                // Block
+                Block.ConsensusData.Size +                      //ConsensusData
+                IO.Helper.GetVarSize(expectedTransactions + 1); //Transactions count
+
+            return blockSize;
+        }
+
+        /// <summary>
+        /// Prevent that block exceed the max size
+        /// </summary>
+        /// <param name="txs">Ordered transactions</param>
+        internal void EnsureMaxBlockSize(IEnumerable<Transaction> txs)
+        {
+            uint maxBlockSize = NativeContract.Policy.GetMaxBlockSize(Snapshot);
+            uint maxTransactionsPerBlock = NativeContract.Policy.GetMaxTransactionsPerBlock(Snapshot);
+
+            // Limit Speaker proposal to the limit `MaxTransactionsPerBlock` or all available transactions of the mempool
+            txs = txs.Take((int)maxTransactionsPerBlock);
+            List<UInt256> hashes = new List<UInt256>();
+            Transactions = new Dictionary<UInt256, Transaction>();
+            Block.Transactions = new Transaction[0];
+
+            // We need to know the expected block size
+
+            var blockSize = GetExpectedBlockSizeWithoutTransactions(txs.Count());
+
+            // Iterate transaction until reach the size
+
+            foreach (Transaction tx in txs)
+            {
+                // Check if maximum block size has been already exceeded with the current selected set
+                blockSize += tx.Size;
+                if (blockSize > maxBlockSize) break;
+
+                hashes.Add(tx.Hash);
+                Transactions.Add(tx.Hash, tx);
+            }
+
+            TransactionHashes = hashes.ToArray();
         }
 
         public ConsensusPayload MakePrepareRequest()
         {
             byte[] buffer = new byte[sizeof(ulong)];
             random.NextBytes(buffer);
-            List<Transaction> transactions = Blockchain.Singleton.MemPool.GetSortedVerifiedTransactions()
-                .Take((int)NativeContract.Policy.GetMaxTransactionsPerBlock(Snapshot))
-                .ToList();
-            TransactionHashes = transactions.Select(p => p.Hash).ToArray();
-            Transactions = transactions.ToDictionary(p => p.Hash);
-            Block.Timestamp = Math.Max(TimeProvider.Current.UtcNow.ToTimestampMS(), PrevHeader.Timestamp + 1);
             Block.ConsensusData.Nonce = BitConverter.ToUInt64(buffer, 0);
+            EnsureMaxBlockSize(Blockchain.Singleton.MemPool.GetSortedVerifiedTransactions());
```

Comment:
Ok, last comment, isn't it better to already ensure this inside GetSortedVerifiedTx? Or isn't it trusted?

Reply:
I agree, but this method GetSorted already ensures that, no @igormcoelho?

Isn't it just calling the function that has this set already sorted? If not, I would prefer as you said, brother.

Reply:
For me one thing is the mempool, and other what i do with these transactions, `EnsureMaxBlockSize` require the `Validators` and this information is here

Reply:
But `Validators` info could also be made available there... or is it only processing Verifications, we count the votes and know next consensus?
Anyway, this is less critical, just a comment, why we don't trust our own mempool module :joy: